### PR TITLE
Effect of $search on RecursiveHierarchy

### PR DESCRIPTION
**Pull request by HeikoTheissen**
This PR provides SAP-specific information about recursive hierarchies in additional entity type properties. This only works under the assumption that the recursive hierarchy is defined on the requested resource itself.

For the more general case, where the hierarchy-carrying entity type can differ from the requested entity type, instance annotations must be used instead of properties, see #184.

### CODE REVIEW

**Review by ralfhandl — APPROVED** [review]

**Review by GeraldKrause — APPROVED** [review]

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:24]: Not related to this change, actually. This is the single mandatory property here and I wonder what it is really needed for? What is its role and where can I see some example or so?

File: vocabularies/Hierarchy.xml
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.xml:53]: Core.LongDescription lost here?

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:26]: I wonder a bit about this enum-like string. Looks strange to me. Why send such long strings on the wire? Human readability in a response targeted at a machine?

"Retrieving them requires a relaxed filter expression or a separate request filtering on the parent node ID with the ID of the collapsed node." Is very technical. Not sure if the advice is really spot on. Maybe less is more here?

Comment (ThomasChadzelek) [reply vocabularies/Hierarchy.md:26]: "result collection" (here) vs. "output set" (line 16ff.)?

"entity set" vs. "input collection" or so - I guess this wording is also not aligned with $apply

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:27]: Does "preorder" matter here?

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:30]: "descending the hierarchy" is kind of an ad-hoc term. Maybe "a child's level is strictly larger than the level of its parent"? And maybe even "it is possible that root nodes are not on level 1 and direct children may be more than one level below their parents" where the 2nd part is a bit redundant, but emphasizes what might be astonishing (copied from https://help.sap.com/doc/e4fc5d26a38c4dc28802a3b87cd5eb10/2.0.05/en-US/SAP_HANA_Hierarchy_Developer_Guide_en.pdf).

File: vocabularies/Hierarchy.xml
Comment (HeikoTheissen) [reply vocabularies/Hierarchy.xml:59]: There is no annotation `Validation.AllowedTypes` that would make this machine-readable.

**Comment by HeikoTheissen** [comment]
> Sorry, but what is the enum "expanded, collapsed, leaf" good for if I cannot distinguish collapsed nodes from leaves? Does that really lead to a good UX?

If the client requests descendants up to level 3, the server may not want to check whether the nodes on level 3 are actually leaves, because that's too expensive. So it would return level 3 nodes as `collapsed`, and higher level ($<3$) nodes as `expanded` or `leaf`. Only when the user clicks on a level 3 node do they learn whether it is a leaf.

I consider that a meaningful use case and do not want to exclude it in this term definition.

**Comment by ThomasChadzelek** [comment]
OK, only time will tell. Or Markus Königstein, paying attention to "V2 compatibility" - do you have an eye on that?

**Comment by HeikoTheissen** [comment]
Of course our servers can promise never to return `collapsed` for leaves, if they feel they can ensure that.

**Comment by ThomasChadzelek** [comment]
How does this relate to IsLeafProperty? If I $select that, I would get the exact information about the requested resource, independent of the output set? That would be nice, the client can then determine what it needs and how much it is willing to pay.

**Comment by HeikoTheissen** [comment]
> How does this relate to IsLeafProperty? If I $select that, I would get the exact information about the requested resource, independent of the output set? That would be nice, the client can then determine what it needs and how much it is willing to pay.

Yes, and the server can determine which of these properties to offer.

**Review by ThomasChadzelek — DISMISSED** [review]

**Review by ralfhandl — DISMISSED** [review]

File: vocabularies/Hierarchy.md
Comment (GeraldKrause) [inline vocabularies/Hierarchy.md:18]:
```suggestion
          and hierarchy-related transformations (`ancestors`, `descendants`, `traverse`) applied to it using the resource as first parameter and the hierarchy qualifier of the annotation as second parameter,
```

I read your description as if you apply the simplification/restriction we discussed in our call: The first parameter of ancestors/descendants is identical the input set. If that's what you intended I would state it clearly as suggested

**Review by ThomasChadzelek — APPROVED** [review]
Who am I to judge?

**Review by ralfhandl — APPROVED** [review]

**Comment by HeikoTheissen** [comment]
> Einspruch, Euer Ehren! Das war bisher für mich die zentrale technische Property, die ich zum Blättern über die "_limited_ hierarchy" brauche! Welcher Ersatz wäre denn dafür vorgesehen?

[`ExpandedDescendantCountProperty`](https://github.com/SAP/odata-vocabularies/pull/182/commits/9e8246ea4e89ad1cb9e6ab2ea5d4272a36a00f72)

File: vocabularies/Hierarchy.xml
Comment (HeikoTheissen) [inline vocabularies/Hierarchy.xml:72]: These definitions, on which the properties of `RecursiveHierarchy` are based, are somewhat fluffy and reflect the thinking of OData V2 hierarchies, where the following kinds of `$filter` were supported:
- filter on node ID: determines a sub-hierarchy
- filter on parent ID: determines children of a node
- filter on `hierarchy_level`: limits the hierarchy
- filter on other property: determines matching nodes.

These are also the kinds of filter that the underlying database supports. OData V4 goes conceptually beyond that, but then the properties of `RecursiveHierarchy` no longer work.

Is that restriction OK?

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:66]: There is no restriction w.r.t. a type here (e.g. Edm.String), right? Would that make sense?

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:65]: There is no restriction w.r.t. a type here (e.g. Edm.String), right? Would that make sense?

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:70]: Maybe we could be less strict here and only request strings that start with "C", "E", or "L" (disregarding case)? Or even use a number, e.g.
- 0 for a leaf
- +1 for a collapsed node ("+" looks like the icon where you could expand)
- -1 for an expanded node (the "opposite" of collapsed)
But maybe that is a bit far fetched?

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [inline vocabularies/Hierarchy.md:72]: Should we add a hint that updating does not make sense when an orderby is present which (of course) influences sibling order? That is, only w/o orderby you get the "native" order which you may be able to PATCH.

By the way: if "unlimited hierarchy" is relevant (as opposed to "(sub-)hierarchy"), then how would that be identified in a PATCH request which targets a single entity directly, w/o $apply?

File: vocabularies/Hierarchy.md
Comment (HeikoTheissen) [reply vocabularies/Hierarchy.md:70]: To be discussed in V4 tree meeting.

File: vocabularies/Hierarchy.md
Comment (ThomasChadzelek) [reply vocabularies/Hierarchy.md:66]: For the node type, I would expect the qualified name of the entity type. Or is that up to the service to even use some number or enum?

For the external key value, I guess "external" means that you can display it, and then a string makes most sense to me. Or could that be even a timestamp and then you need to format it incl. all dicussions about timezone?

Comment (HeikoTheissen) [reply vocabularies/Hierarchy.md:66]: The external key may have a complex type made up from the node type and a node-specific id.

File: vocabularies/Hierarchy.xml
Comment (HeikoTheissen) [inline vocabularies/Hierarchy.xml:138]:
```suggestion
          This subset is the output set of the `filter` or `search` transformation in the fourth parameter of the last `ancestors` transformation.
          &lt;br&gt;This instance annotation is available if [`RecursiveHierarchy/MatchedProperty`](#RecursiveHierarchyType)
```
Would this be equivalent for our purposes? It is certainly easier to implement.

Comment (ThomasChadzelek) [reply vocabularies/Hierarchy.xml:138]: Looks good to me. I agree that "ancestors" is the relevant transformation for this kind of search/filter: you need to locate matches and the complete the tree "upwards". A "descendants" transformation has a different semantics, it restricts to a sub-tree, but that is not what a "match" needs to indicate. You will notice that effect by looking at the resulting root node(s).

**Review by ThomasChadzelek — APPROVED** [review]

**Comment by HeikoTheissen** [comment]
We plan to merge this PR at 2023-02-24T15:00:00+01:00 (after the V4 tree meeting). Until then
- we need at least one approving review by someone with write access to this repository
- please raise any objections you might have with the current proposal especially of the `Hierarchy.RecursiveHierarchy` annotation type.

**Review by ThomasChadzelek — DISMISSED** [review]

File: vocabularies/Hierarchy.xml
Comment (ralfhandl) [inline vocabularies/Hierarchy.xml:94]:
```suggestion
            &lt;br&gt;- be represented by entities of different subtypes of a common entity type that is
```

**Review by ralfhandl — CHANGES_REQUESTED** [review]
It's a bit surprising that all properties of `RecursiveHierarchyType` are nullable.

**Comment by HeikoTheissen** [comment]
> It's a bit surprising that all properties of `RecursiveHierarchyType` are nullable.

None of them is so important that it must always be present.

**Comment by ralfhandl** [comment]
> None of them is so important that it must always be present.

Is there a use case where none of the properties is present and the annotation only serves as a marker?

**Review by ralfhandl — DISMISSED** [review]

**Comment by ThomasChadzelek** [comment]
> > Is there a use case where none of the properties is present and the annotation only serves as a marker?
> 

From the perspective of UI5's v4.ODataModel, DrillStateProperty is an absolute must. If you don't know whether a node is a leaf or not, you are lost.

LimitedDescendantCountProperty and DistanceFromRootProperty are mandatory as soon as you want to start with n > 1 initially expanded levels, which is the reason we invented TopLevels and the main benefit of V4 over V2, right?

The rest is optional ("graceful degradation")

**Comment by HeikoTheissen** [comment]
> From the perspective of UI5's v4.ODataModel, DrillStateProperty is an absolute must. If you don't know whether a node is a leaf or not, you are lost.
> LimitedDescendantCountProperty and DistanceFromRootProperty are mandatory as soon as you want to start with n > 1 initially expanded levels, which is the reason we invented TopLevels and the main benefit of V4 over V2, right?

But this annotation is meant for all future UI frameworks at SAP, which may not have the same needs as UI5.

**Comment by ralfhandl** [comment]
> But this annotation is meant for all future UI frameworks at SAP, which may not have the same needs as UI5.

How about starting with non-nullable DrillStateProperty and relax this when the first UI framework arrives that doesn't require it?

**Review by ralfhandl — APPROVED** [review]

**Pull request merged**
